llvm/test/tools/llvm-cov/native_separators.c

```diff
@@ -3,6 +3,7 @@
 // This test is Windows-only. It checks that all paths, which are generated
 // in the index and source coverage reports, are native path. For example,
 // on Windows all '/' are converted to '\'.
+// UNSUPPORTED: system-windows
```

Comment:
This and the following lines are an odd combination, especially give the opening comment.  More generally, I would have liked comments on at least some of the Windows Unsupported directives.  Even more generally, some indication of which are because of LLVM’s relative lack of interest in Windows, and which because of Windows’ many limitations and nonconformities.  (Both can be true, of course.)

Reply:
@FlashSheridan , thanks a lot for a thorough review. You're absolutely right that we need to categorize and add more comments in all of these cases.

I actually wanted to separate these issues into categories and understand on how to proceed further with them.

LLVM issues can be split into the following categories:

1. Windows MSYS CI environment returns `BAD-DATE-FORMAT` instead of the date and it fails the following tests that checks the date in output:
`LLVM :: Object/archive-big-read.test`
`LLVM :: Object/archive-toc.test`
`LLVM :: Object/archive-extract.test`
`LLVM :: tools/llvm-ar/empty-uid-gid.test`
`LLVM :: tools/llvm-libtool-darwin/deterministic-library.test`
`LLVM :: tools/llvm-objcopy/ELF/deterministic-archive.test`
`LLVM :: tools/llvm-ranlib/D-flag.test`

2. "Unsupported file type" string returned in our Windows MSYS environment instead of "malformed file" that is expected:
`LLVM :: tools/llvm-nm/invalid-tapi-files.test`
`LLVM :: tools/llvm-nm/tapi-files.test`
`LLVM :: tools/llvm-tapi-diff/v5.test`

3. Directory and Linux/Windows slash `/``\` issues:
* `LLVM :: TableGen/x86-fold-tables.td` - Linux specific slashes in test, not working on Windows properly
* `LLVM :: tools/llvm-cov/native_separators.c` - Windows-specific test, but it will never work in msys environment as it tries to use \tmp directory. It will only work on Windows native.
* `LLVM :: tools/llvm-cov/relative-dir.test` - multiple slashes on MSys are generated and leads to wrong output check
* `LLVM-Unit :: Support/./SupportTests.exe/55/83` - path issue that is impossible to debug without native Windows environment

I'm not sure what to do with that, it's impossible to debug without native and MSYS environments together, and this can be some bug in the LLVM Support Chrono library. These tests are also not really relevant for us, so I believe it's probably OK to disable them completely marking them as `UNSUPPORTED` on Windows and adding a comment about it. What do you think?

The next thing is the LLD tests:

1. Require proper investigation when we have Windows native host:
* `LLD-Unit :: AsLibELF/./LLDAsLibELFTests.exe/1/2` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eravm-code-reloc.s` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eravm-data-reloc.s`- Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eravm-errors.s` - Some additional output "ld.lld: warning: cannot find entry symbol _start; not setting start address" that is not expected by the test
* `lld :: ELF/invalid/symtab-symbols.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/eh-frame-type.test` - Non zero exit code returned by ld.lld
* `lld :: ELF/llvm33-rela-outside-group.s` - Exception fail in llvm::PassRegistry::~PassRegistry()
* `lld :: ELF/relocation-group.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/relocation-none.test `- Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/section-align-0.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/reloc-sec-before-relocated.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/relocation-rel-format.test` - Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/sht-group-empty.test `- Non zero exit code returned by ld.lld without any stderr output
* `lld :: ELF/shf-info-link.test` - Non zero exit code returned by ld.lld without any stderr output

2. MacOS specific tests that are unsupported for Windows by default:
`lld :: MachO/error-limit.test`
`lld :: MachO/double-unwind-info.s`
`lld :: MachO/local-private-extern.yaml`

For these issues, I've created [CPR-1784](https://linear.app/matterlabs/issue/CPR-1784/llvm-check-lld-regression-tests-windows-failures) and they will require investigation as soon as we have Windows host.

I was also thinking to disable them for now, because I was unable to make them to `--xfail` properly, it's just doesn't work on Windows :(

It's a huge pain to investigate and debug these issues without a Windows host.

Reply:
_(1 of 2)_
> It's a huge pain to investigate and debug these issues without a Windows host.

Of course; supporting Windows (or any platform) has an absolute requirement for at least one machine (perhaps virtual) to test it on; I’ve made that clear in planning documents since before you started at Matter Labs.  I don’t think it’s a good use of your time to waste it on unnecessarily difficult experimentation while waiting for required hardware.  If you need backing for escalating this, count me in, though given the cheapness of Windows hardware, it seems to be covered by Anthony’s policy:

> 2024-03-22 Here is the new policy for the tech tribe: if there is a time sensitive expense for <$1000, the team leads can just approve it without checking with me or anyone else. … and if it was something that we shouldn't have approved, we can discuss and learn from this (but will still reimburse). 

Reply:
_(2 of 2)_

> You're absolutely right that we need to categorize and add more comments in all of these cases.
> I actually wanted to separate these issues into categories and understand on how to proceed further with them.
> LLVM issues can be split into the following categories: …

This is excellent work and deserves a wider audience, both internal and upstream, even if it exceeds in both length and quality what’s customary for LLVM’s Windows-specific documentation and comments.  If I were working on another LLVM fork and struggling with Windows support, I’d be delighted to see your comments (at the latest when we upstream our fork, perhaps earlier) for each suppression, the information you’ve elicited on whether the suppression is because of:
1.  a genuine Windows limitation, 
2. legitimate target-specificity in the test
3. laziness in the test, 
4. a genuine bug in the upstream software under test, 
5. a (preferably temporary) limitation in our code, or
6. currently unknown.